### ISSUE TITLE
MPI_Status OUT LIS is incorrect

### ISSUE THREAD

**Issue by martinruefenacht** [mpi-forum/mpi-issues#516]
# Problem

The status parameters of all procedures which use a MPI_Status with an "OUT" direction have an incorrect LIS "direction".

The definition of OUT is: `the call may update the argument but does not use its input value`. The status parameter is used to determine the action on the status parameter. If MPI_STATUS_IGNORE is passed (read/used by the implementation), then the parameter is not used/updated. This fits the INOUT definition: `the call may both use and update the argument`.

This is equally applicable to the "array_of_statuses" parameters. Affected procedures are:

OUT status:
- MPI_Recv
- MPI_Sendrecv
- MPI_Sendrecv_replace
- MPI_Wait
- MPI_Test
- MPI_Waitany
- MPI_Testany
- MPI_Request_get_status
- MPI_Iprobe
- MPI_Probe
- MPI_Improbe
- MPI_Mprobe
- MPI_Mrecv
- MPI_File_read_at
- MPI_File_read_at_all
- MPI_File_write_at
- MPI_File_write_at_all
- MPI_File_read
- MPI_File_read_all
- MPI_File_write
- MPI_File_write_all
- MPI_File_read_shared
- MPI_File_write_shared
- MPI_File_read_ordered
- MPI_File_write_ordered
- MPI_File_read_at_all_end
- MPI_File_write_at_all_end
- MPI_File_read_all_end
- MPI_File_write_all_end
- MPI_File_read_ordered_end
- MPI_File_write_ordered_end

OUT array_of_statuses:
- MPI_Waitall
- MPI_Testall
- MPI_Waitsome
- MPI_Testsome

Correctly specified (already INOUT):
- MPI_Status_set_elements
- MPI_Status_set_elements_x
- MPI_Status_set_cancelled

This change is already broadly in place due to a sentence on page 41 of the MPI Standard 4.0.
`Note that this converts status into an INOUT argument.` This sentence is not reflected in the LIS definitions of any of the affected procedures. And this sentence does not cover the entire complexity for all procedures.

This does not affect MPI_Test_cancelled, MPI_Get_elements, MPI_Get_elements_x. The status parameter in marked as IN. Since it must be readable MPI_STATUS_IGNORE is not valid. (A change to be addressed separately.)

# Proposal

For the affected procedures we need to change the OUT of the status parameter to INOUT for the LIS only. The language expressions are already appropriate for the languages.

# Changes to the Text

Change appropriate status LIS descriptors to be INOUT.

# Impact on Implementations

None.

# Impact on Users

None.

# References and Pull Requests

MPI 4.0, page 12, Procedure Specification, OUT/INOUT definitions
MPI 4.0, page 41, 3.2.6, Passing MPI_STATUS_IGNORE for Status

**Comment by RolfRabenseifner** [comment]
MPI-3.1 page 10 line 37 and MPI-4.0 page 12 line 34 clearly shows that passing such special addresses like MPI_BOTTOM or MPI_STATUS_IGNORE also to OUT arguments. The OUT refers to the writing of bits of content. Consequently, the IN marks the reading of bits of content. Instead of changing the whole meaning of IN and OUT in the MPI standard, one may add one sentence to Section 2.3.  We have been very careful in Fortran, i.e., not defining some INTENT, because the implementation must be able to rely on call by reference (instead of any IN-copy or OUT-copy) or IN-and-OUT-copy).

I hope, that there is no hidden information about the syntax or semantics of an MPI API in the current latex/python source.
Hidden in this sentence means, that it may be taken for any other purpose, but is not under control of the MPI forum when reading the MPI standard pdf.

(There is such hidden info, e.g., all formatting, the additional infos for checking whether examples can be compiled. But they are not part of the API.)

**Comment by martinruefenacht** [comment]
Agreed, the sentence permits MPI_STATUS_IGNORE to be passed to a OUT parameter. I do not mean to change the meaning of IN or OUT since those are *well defined* on page 12. In those definitions the word *use* is used, not reading or writing.

@Wee-Free-Scot and I did have a discussion about meaning of use, is reading included in that definition or is that a separate *action*.

The contention is not about the OUT component, but about the IN component. I am contending that the MPI implementation must read(use) the bits(value) of the status parameter to be able to determine whether it should write bits out to the parameter.

You mention call by reference with which I understand pass-by-reference? The method which was chosen for status actually forces both pass-by-reference (writing out to an MPI_Status) and pass-by-value(checking for MPI_STATUS_IGNORE).

The sentences in section 3.2.6: `In general, this optimization can apply to all functions for which status or an array of
statuses is an OUT argument. Note that this converts status into an INOUT argument.`, contradict the sentence you pointed out.

Could you write the one sentence you mean?

The Python only contains an IN, OUT, INOUT enum so no hidden information.

**Comment by RolfRabenseifner** [comment]
I recommend to clearly differentiate between the address of an argument and the content of the memory this address is pointing to if the argument is a pass-by-reference argument. The an OUT status argument is always pass-by-reference. It is a reference to a structure. MPI_STATUS_IGNORE is a special address and not a reference to a structure with a special value. If the user passes a structure with any (undefined) content then it must be interpreted by the MPI routine always as not being MPI_STATUS_IGNORE.
The list on MPI-4.0 page 21 starting with MPI_BOTTOM shows you all such special references. 
E.g., for MPI_BOTTOM, since MPI-1, all MPI receive functions do not specify their buffer argument as INOUT.
I expect that it was clear for the past 25 years. If not then Section 2.3 may need an additional sentence after examination of the whole textr of 2.3. I'm fine with the text we have now for many years.

**Comment by RolfRabenseifner** [comment]
And same with passing MPI_IN_PLACE to the OUT recvbuf argument in MPI_Scatter.

A sentence can be: 
> For the distinguishing of IN or OUT it is not taken into account whether it is possibe to pass the special constants listed on page 21 starting with MPI_BOTTOM.

But I do not believe that this is needed.
